// @progress/kendo-angular-upload/dist/npm/rendering/file-list.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
var types_1 = require("../types");
var file_template_directive_1 = require("../templates/file-template.directive");
var file_list_item_1 = require("./file-list-item");
var navigation_service_1 = require("../navigation.service");
var upload_service_1 = require("../upload.service");
/**
 * @hidden
 */
var FileListComponent = /** @class */ (function () {
    function FileListComponent(uploadService, navigation) {
        this.uploadService = uploadService;
        this.navigation = navigation;
        this.onItemFocus();
        this.onItemAction();
    }
    FileListComponent.prototype.onItemFocus = function () {
        var _this = this;
        this.focusSubscription = this.navigation.onFileFocus.subscribe(function (index) {
            _this.fileListItems.toArray()[index].focus();
        });
    };
    FileListComponent.prototype.onItemAction = function () {
        var _this = this;
        this.actionSubscription = this.navigation.onFileAction.subscribe(function (key) {
            _this.itemActionHandler(key);
        });
    };
    FileListComponent.prototype.itemActionHandler = function (key) {
        var index = this.navigation.focusedIndex;
        var item = this.fileListItems.toArray()[index];
        var uid = item.uidAttribute;
        var files = this.uploadService.files.get(uid);
        if (key === kendo_angular_common_1.Keys.Escape && files[0].state === types_1.FileState.Uploading) {
            this.uploadService.cancelFiles(uid);
            this.navigation.focusSelectButton();
            return;
        }
        if (key === kendo_angular_common_1.Keys.Enter && files[0].state === types_1.FileState.Failed) {
            this.uploadService.retryFiles(uid);
            return;
        }
        if (key === kendo_angular_common_1.Keys.Delete) {
            if (files[0].state === types_1.FileState.Uploading) {
                this.uploadService.cancelFiles(uid);
            }
            else if (this.hasDelete(item)) {
                this.uploadService.removeFiles(uid);
            }
            this.navigation.focusSelectButton();
        }
    };
    FileListComponent.prototype.hasDelete = function (item) {
        return item.element.nativeElement.getElementsByClassName('k-delete').length > 0;
    };
    FileListComponent.prototype.ngOnDestroy = function () {
        this.focusSubscription.unsubscribe();
        this.actionSubscription.unsubscribe();
    };
    FileListComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: '[kendo-upload-file-list]',
                    template: "\n    <ng-template ngFor\n      [ngForOf]=\"fileList\"\n      let-files\n      let-index=\"index\">\n      <li kendoUploadFileListItem [files]='files' [index]='index'>\n          <kendo-upload-file-list-single-item\n            class='k-file-single'\n            *ngIf='files.length === 1 && !fileTemplate'\n            [disabled]='disabled'\n            [file]='files[0]'>\n          </kendo-upload-file-list-single-item>\n          <kendo-upload-file-list-multiple-items\n            class='k-file-multiple'\n            *ngIf='files.length > 1 && !fileTemplate'\n            [disabled]='disabled'\n            [files]='files'>\n          </kendo-upload-file-list-multiple-items>\n          <ng-template *ngIf=\"fileTemplate\"\n              [templateContext]=\"{\n                templateRef: fileTemplate.templateRef,\n                state: files[0].state,\n                $implicit: files\n              }\"></ng-template>\n      </li>\n    </ng-template>\n    "
                },] },
    ];
    /** @nocollapse */
    FileListComponent.ctorParameters = function () { return [
        { type: upload_service_1.UploadService },
        { type: navigation_service_1.NavigationService }
    ]; };
    FileListComponent.propDecorators = {
        disabled: [{ type: core_1.Input }],
        fileList: [{ type: core_1.Input }],
        fileTemplate: [{ type: core_1.Input }],
        fileListItems: [{ type: core_1.ViewChildren, args: [file_list_item_1.FileListItemDirective,] }]
    };
    return FileListComponent;
}());
exports.FileListComponent = FileListComponent;
