// @progress/kendo-angular-upload/dist/npm/navigation.service.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var upload_service_1 = require("./upload.service");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
/**
 * @hidden
 */
var NavigationService = /** @class */ (function () {
    function NavigationService(uploadService) {
        this.uploadService = uploadService;
        this.onActionButtonAction = new core_1.EventEmitter();
        this.onActionButtonFocus = new core_1.EventEmitter();
        this.onFileAction = new core_1.EventEmitter();
        this.onFileFocus = new core_1.EventEmitter();
        this.onTab = new core_1.EventEmitter();
        this.onWrapperFocus = new core_1.EventEmitter();
        this.onSelectButtonFocus = new core_1.EventEmitter();
        this.actionButtonsVisible = false;
        this.focused = false;
        this._focusedIndex = -1;
    }
    NavigationService.prototype.action = function (event) {
        var key = event.keyCode;
        return this.keyBindings[key];
    };
    NavigationService.prototype.process = function (event) {
        var handler = this.action(event);
        if (handler) {
            handler(event.shiftKey);
        }
    };
    NavigationService.prototype.computeKeys = function (direction) {
        var _this = this;
        var _a;
        this.keyBindings = (_a = {},
            _a[kendo_angular_common_1.Keys.Enter] = function () { return _this.handleEnter(); },
            _a[kendo_angular_common_1.Keys.Escape] = function () { return _this.handleEscape(); },
            _a[kendo_angular_common_1.Keys.Delete] = function () { return _this.handleDelete(); },
            _a[kendo_angular_common_1.Keys.Tab] = function (shifted) { return _this.handleTab(shifted); },
            _a[kendo_angular_common_1.Keys.ArrowUp] = function () { return _this.handleUp(); },
            _a[kendo_angular_common_1.Keys.ArrowDown] = function () { return _this.handleDown(); },
            _a[this.invertKeys(direction, kendo_angular_common_1.Keys.ArrowLeft, kendo_angular_common_1.Keys.ArrowRight)] = function () { return _this.handleLeft(); },
            _a[this.invertKeys(direction, kendo_angular_common_1.Keys.ArrowRight, kendo_angular_common_1.Keys.ArrowLeft)] = function () { return _this.handleRight(); },
            _a);
    };
    NavigationService.prototype.invertKeys = function (direction, original, inverted) {
        return direction === 'rtl' ? inverted : original;
    };
    NavigationService.prototype.focusSelectButton = function () {
        this.focused = true;
        this._focusedIndex = -1;
        this.onSelectButtonFocus.emit();
    };
    NavigationService.prototype.handleEnter = function () {
        if (this.lastIndex >= 0) {
            if (this.focusedIndex <= this.lastFileIndex) {
                this.onFileAction.emit(kendo_angular_common_1.Keys.Enter);
                return;
            }
            if (this.actionButtonsVisible && this.focusedIndex <= this.lastIndex) {
                this.onActionButtonAction.emit(this.focusedIndex < this.lastIndex ? "clear" : "upload");
            }
        }
    };
    NavigationService.prototype.handleDelete = function () {
        if (this.focusedIndex >= 0 && this.focusedIndex <= this.lastFileIndex) {
            this.onFileAction.emit(kendo_angular_common_1.Keys.Delete);
        }
    };
    NavigationService.prototype.handleEscape = function () {
        if (this.focusedIndex >= 0 && this.focusedIndex <= this.lastFileIndex) {
            this.onFileAction.emit(kendo_angular_common_1.Keys.Escape);
        }
    };
    NavigationService.prototype.handleLeft = function () {
        if (this.actionButtonsVisible && this.focusedIndex === this.lastIndex) {
            this.focusedIndex -= 1;
            this.onActionButtonFocus.emit("clear");
        }
    };
    NavigationService.prototype.handleRight = function () {
        if (this.actionButtonsVisible && this.focusedIndex === this.lastIndex - 1) {
            this.focusedIndex += 1;
            this.onActionButtonFocus.emit("upload");
        }
    };
    NavigationService.prototype.handleTab = function (shifted) {
        if (this.focusedIndex >= 0 && shifted) {
            this.focusedIndex = -1;
            return;
        }
        this.onTab.emit();
    };
    NavigationService.prototype.handleDown = function () {
        if (this.lastIndex >= 0 && this.focusedIndex < this.lastIndex) {
            if (this.focusedIndex < this.lastFileIndex) {
                this.focusedIndex += 1;
                this.onFileFocus.emit(this.focusedIndex);
                return;
            }
            if (this.actionButtonsVisible && this.focusedIndex === this.lastFileIndex) {
                this.focusedIndex += 1;
                this.onActionButtonFocus.emit("clear");
            }
        }
    };
    NavigationService.prototype.handleUp = function () {
        if (this.lastIndex >= 0 && this.focusedIndex > -1) {
            this.focusedIndex -= 1;
            if (this.focusedIndex === -1) {
                this.onSelectButtonFocus.emit();
                return;
            }
            if (this.focusedIndex <= this.lastFileIndex) {
                this.onFileFocus.emit(this.focusedIndex);
                return;
            }
            if (this.actionButtonsVisible && this.focusedIndex <= this.lastIndex) {
                this.focusedIndex = this.lastFileIndex;
                this.onFileFocus.emit(this.focusedIndex);
            }
        }
    };
    Object.defineProperty(NavigationService.prototype, "focusedIndex", {
        get: function () {
            return this._focusedIndex;
        },
        set: function (index) {
            if (!this.focused) {
                this.onWrapperFocus.emit();
            }
            this._focusedIndex = index;
            this.focused = true;
            if (this._focusedIndex >= 0 && this._focusedIndex <= this.lastFileIndex) {
                this.onFileFocus.emit(index);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationService.prototype, "lastFileIndex", {
        get: function () {
            return this.actionButtonsVisible ? this.lastIndex - 2 : this.lastIndex;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavigationService.prototype, "lastIndex", {
        get: function () {
            var fileCount = this.uploadService.files.count;
            return this.actionButtonsVisible ? fileCount + 1 : fileCount - 1;
        },
        enumerable: true,
        configurable: true
    });
    NavigationService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    NavigationService.ctorParameters = function () { return [
        { type: upload_service_1.UploadService }
    ]; };
    return NavigationService;
}());
exports.NavigationService = NavigationService;
