// @progress/kendo-angular-grid/dist/npm/navigation/logical-row.directive.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2020 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var id_service_1 = require("../common/id.service");
var navigation_service_1 = require("./navigation.service");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
var id = 0;
function nextId() {
    return id++;
}
/**
 * @hidden
 */
var LogicalRowDirective = /** @class */ (function () {
    function LogicalRowDirective(idService, navigation) {
        this.idService = idService;
        this.navigation = navigation;
        this.logicalSlaveRow = false;
        this.logicalSlaveCellsCount = 0;
        this.dataRowIndex = -1;
        this.uid = nextId();
    }
    Object.defineProperty(LogicalRowDirective.prototype, "hostRole", {
        get: function () {
            return this.logicalSlaveRow ? 'presentation' : 'row';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LogicalRowDirective.prototype, "ariaRowIndex", {
        get: function () {
            if (this.navigation.enabled) {
                return this.logicalRowIndex + 1;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LogicalRowDirective.prototype, "ariaOwns", {
        get: function () {
            if (!this.navigation.enabled || this.logicalSlaveRow || this.logicalSlaveCellsCount === 0) {
                return undefined;
            }
            var ids = [];
            var total = this.logicalCellsCount + this.logicalSlaveCellsCount;
            for (var cellIndex = this.logicalCellsCount; cellIndex < total; cellIndex++) {
                ids.push(this.idService.cellId(this.logicalRowIndex, cellIndex));
            }
            return ids.join(' ');
        },
        enumerable: true,
        configurable: true
    });
    LogicalRowDirective.prototype.ngOnChanges = function (changes) {
        if (!this.navigation.enabled || this.logicalSlaveRow) {
            return;
        }
        var indexChange = changes.logicalRowIndex;
        var logicalSlaveRowChange = changes.logicalSlaveRow;
        if (indexChange || logicalSlaveRowChange) {
            var index = indexChange && !indexChange.isFirstChange() ? indexChange.previousValue : this.logicalRowIndex;
            this.navigation.unregisterRow(index, this);
            this.navigation.registerRow(this);
        }
        else if (kendo_angular_common_1.anyChanged(['dataRowIndex', 'dataItem'], changes)) {
            this.navigation.updateRow(this);
        }
    };
    LogicalRowDirective.prototype.ngOnDestroy = function () {
        this.navigation.unregisterRow(this.logicalRowIndex, this);
    };
    LogicalRowDirective.decorators = [
        { type: core_1.Directive, args: [{
                    selector: '[kendoGridLogicalRow]'
                },] },
    ];
    /** @nocollapse */
    LogicalRowDirective.ctorParameters = function () { return [
        { type: id_service_1.IdService },
        { type: navigation_service_1.NavigationService }
    ]; };
    LogicalRowDirective.propDecorators = {
        logicalRowIndex: [{ type: core_1.Input }],
        logicalSlaveRow: [{ type: core_1.Input }],
        logicalCellsCount: [{ type: core_1.Input }],
        logicalSlaveCellsCount: [{ type: core_1.Input }],
        dataRowIndex: [{ type: core_1.Input }],
        dataItem: [{ type: core_1.Input }],
        hostRole: [{ type: core_1.HostBinding, args: ['attr.role',] }],
        ariaRowIndex: [{ type: core_1.HostBinding, args: ['attr.aria-rowindex',] }],
        ariaOwns: [{ type: core_1.HostBinding, args: ['attr.aria-owns',] }]
    };
    return LogicalRowDirective;
}());
exports.LogicalRowDirective = LogicalRowDirective;
